rollup.config.js

```diff
@@ -3,41 +3,150 @@ import nodeResolve from 'rollup-plugin-node-resolve';
 import babel from 'rollup-plugin-babel';
 import replace from 'rollup-plugin-replace';
 import uglify from 'rollup-plugin-uglify';
+import pkg from './package.json';
 
-const env = process.env.NODE_ENV
-const config = {
-  output: {
-    format: 'umd',
-    name: 'ReduxPersist',
-    exports: 'named',
-    sourcemap: true,  
-  },
+// const ensureArray = maybeArr => Array.isArray(maybeArr) ? maybeArr : [maybeArr];
+
+// const createConfig = ({
+//   input = 'src/index.js',
+//   output,
+//   umd = false,
+//   env,
+// } = {}) => ({
+//   input,
+//   output: ensureArray(output).map(format => Object.assign(
+//     {},
+//     format,
+//     umd
+//       ? {
+//         name: 'ReduxPersist',
+//         sourcemap: true,
+//       }
+//       : {}
+//   )),
+//   external: umd ? [] : [
+//     ...Object.keys(pkg.dependencies || {}),
+//     ...Object.keys(pkg.peerDependencies || {}),
+//     'react',
+//   ],
+//   plugins: [
+//     nodeResolve({
+//       jsnext: true
+//     }),
+//     babel({
+//       exclude: 'node_modules/**',
+//     }),
+//     umd && replace({
+//       'process.env.NODE_ENV': JSON.stringify(env),
+//     }),
+//     umd && env === 'production' && uglify({
+//       compress: {
+//         pure_getters: true,
+//         unsafe: true,
+//         unsafe_comps: true,
+//         warnings: false,
+//       },
+//     }),
+//   ].filter(Boolean),
+// })
+
+// export default [
+//   createConfig({
+//     output: [
+//       { file: `${pkg.main}.js`, format: 'cjs' },
+//       { file: `${pkg.module}.js`, format: 'es' },
+//     ],
+//   }),
+//   createConfig({
+//     umd: true,
+//     env: 'development',
+//     output: {
+//       file: `${pkg.main}.umd.js`,
+//       format: 'umd',
+//     },
+//   }),
+//   createConfig({
+//     umd: true,
+//     env: 'production',
+//     output: {
+//       file: `${pkg.main}.umd.min.js`,
+//       format: 'umd',
+//     },
+//   }),
+//   createConfig({
+//     input: 'src/integration/react.js',
+//     output: [
+//       { file: `dist/${pkg.name}-react.js`, format: 'cjs' },
+//       { file: `dist/${pkg.name}-react.es.js`, format: 'es' },
+//     ]
+//   })
+// ]
+
+const createConfig = ({
+  input = 'src/index.js',
+  output,
+  umd = false,
+  env,
+} = {}) => ({
+  experimentalCodeSplitting: !umd,
```

Comment:
this version is interesting though, because it uses code splitting feature - this way babel helpers can be deduped between `redux-persist`, `redux-persist/storage` and `redux-persist/react` flat bundles. I've found a bug in rollup while experimenting with this though - https://github.com/rollup/rollup/issues/1943

It would also require restructuring redux-persist src directory structure slightly, because rollup maps original input filenames to output names (not including directories) and we need stable mapping between those

Reply:
cant say I understand. So how would src be restructured?

How would babel helpers get deduped?

Reply:
At the moment with this:
```js
input: ['src/index.js', 'src/integration/react.js', 'src/storage/index.js']
```

It produces 4 files:
- index.js (what will be in main/module)
- index2.js (this should actually be `storage` for readability, that's why we need to restructure a little bit, so input file names get unique names)
- react.js 
- chunk1.js - here we have magically deduped babel helpers :)